File: projects/project_2/resource.rb
class Resource
  attr_accessor :name
  def initialize(name)
    @name = name
    @process = nil
  end

  def process(process)
    @process = process
  end

  def remove_process(process)
    @process = nil
  end

  def process?
    @process ? true : false
  end

  def add_to_queue process
    @queue << process
  end

  def remove_from_queue process
    @queue.remove(process)
  end

  def empty_queue?
    @queue ? false : true
  end

  def pop_queue
    return @queue.shift
  end
end

File: projects/project_2/javascript_solution/public/js/resource.js
/*
* All the functions related to resources
*/

/*
* Create a new resource
*/
function createResource(pName) {
  hash = {};
  hash["name"] = pName;
  hash["waitingProcs"] = [];

  return hash;
}


/*
* Add a process to the waiting queue
*/
function addWaitingProc(resource, process) {
  resource["waitingProcs"].push(process);

  return resource;
}


/*
* Set the owner process of the resource
*/
function setProcess(resource, process) {
  resource["process"] = process;
  return resource;
}


/*
* Remove the owner process And possibly set new process as owner
*/
function removeProcess(resource) {
  if(resource["waitingProcs"]) {
    resource["process"] = resource["waitingProcs"].reverse().pop();
  } else {
    resource["process"] = null;
  }

  return resource;
}

/*
* Check if a resource already exists
*/
function exists(resource, resources) {
  for(var index in resources) {
    if(resources[index]["name"] == resource) {
      return true;
    }
  }
  return false;
}

/*
* Find an exisiting resource
*/
function findResource(resource, resources) {
  for(var index in resources) {
    if(resources[index]["name"] == resource) {
      return resources[index];
    }
  }
  return null;
}
